content/blog/2019-11-13-DynamicEdgeProfiling/index.md

```diff
@@ -0,0 +1,122 @@
++++
+title = "Dynamic Edge Profiling with Optimal Counter Placement"
+[extra]
+bio = """
+  [Katy](https://github.com/kavoor) is a Senior interested in Compilers and Systems Programming. She enjoys playing soccer and exploring Ithaca in her free time.
+  [Henry](https://www.linkedin.com/in/liuhenry4428)
+"""
++++
+
+## Introduction
+
+For this project, we wanted to implement an LLVM pass that dynamically profiled basic block edges. The naïve implementation would be to add a profiling mechanism to every edge in the program, but that comes with a significant amount of overhead. To minimize overhead, we strived to add profiling to the minimum number of edges necessary to obtain the same amount of profiling information. The key insight is that we can determine if an edge is traversed using profiling information from preceding and succeeding edges if the CFG structure ensures the program can only run through a simple path between those profiled edges.
+
+## Algorithm
+To select the edges to profile, we used Knuth's algorithm:
+
+---
+<!-- ![](https://i.imgur.com/NbavLnq.png) -->
+<img src="array.png" style="max-width: 100%" >
+---
+Essentially, this algorithm first finds the (acyclic) minimum spanning tree of the edges in a function and then adds profiling instrumentation---code that logs profiling information---to all edges not in the spanning tree. Note that this algorithm uses the first block in a function as the root of the spanning tree, and that edges are treated as bidirectional when checking for cycles.
+
+This algorithm works because the acyclic nature of the spanning tree entails that there is exactly one path between any two instrumented edges (or the entry or exit of the function). Thus, logging the traversal of one instrumented edge implies traversal of the program through the path from the previously logged edge location to the current one. This also implies that this is the minimum number of instrumented edges because having one less instrumented edge would either create a path that could be traversed without logging any profiling information or create multiple paths that lack sufficient profiling information to deduce which path was taken. 
+
+### Offline Edge Count Extrapolation Algorithm
+After we run our program, we know the number of times our instrumented edges were traversed. In order to determine the number of times the remaining edges were produced, we need to do some post-processing of our CFG with our runtime results.
+
+The idea behind this algorithm is to iterate over the edges and fill in the edge count until convergence. The program will converge under the assumption that the edge profiling information supplied is sufficient to determine the rest of the edge counts. It is valid to fill in an edge count if there is a simple path between two nodes with known edge count. You can find detailed pseudocode and proofs for this offline algorithm in the [LLVM Publication](https://llvm.org/pubs/2010-04-NeustifterProfiling.pdf).
```

Comment:
Instead of just saying "LLVM Publication," maybe refer to the paper by the title or the author? (Actually, it seems to be a dissertation, not a paper.)